### DISCUSSION TITLE
[proposal] Consolidate label management for Issues and PRs

### DISCUSSION THREAD

**Discussion by ekowtaylor** [magma/magma#5370] [Ideas]
Labels provide a nifty way to categorize issues and PRs to make it a lot easier for users of the repo to find things. Label management is currently non-existent and as the community grows, it’s going to be important for is to harness the power of this great feature to help is get better organized.

### Why this is important

* The Magma repo is huge and we want to make it a lot easier for users and developers to find anything
* Magma has a vibrant and growing community working on different aspects of the product and labels provide a means for us to streamline intake for new features, enhancements, bugs etc.

###  Where are we now
* Github provides seven default labels which are very generic and are added to help developers get started. Some of these still exist in the Magma codebase
* We’ve grown our existing [labels](https://github.com/magma/magma/labels) to 76 as at the day this was posted, but many of them remain unused or seldom used
* There has been some effort to clean-up labels, but there is still way to go and this effort is to bridge the gap

### The future of labels

* We’re going to introduce a couple of different types of labels to the Magma repo
* Those types are: 
    * *status labels*
    * *type labels*
    * *priority labels*
    * *custom labels*
* We prefix the name of each label with what type of label it is, so something like “*s**tatus: <Name of Label>*”
* Besides these planned standardized labels, new labels are welcome but lets practice good judgement to ensure that it is absolutely necessary

#### Status labels

* status: abandoned
* status: accepted
* status: available
* status: blocked
* status: completed
* status: in progress
* status: rejected
* status: on hold
* status: pending
* status: planned
* status: review needed
* status: revision needed

#### Type labels

* *Issue type: broad categorization of issues*
    * type: bug
    * type: question
    * type: maintenance
    * type: enhancement
    * type: feature
    * type: proposal
* *Component or service*
    * component: agw
    * component: cwag
    * component: orc8r
    * component: grpc
    * component: mme
    * component: feg
    * component: nms
    * component: docs
    * component: lib
    * component: ci
    * component: sessiond
    * component: teravm
    * component: python
* *Product use-case*
    * product: fwa
    * product: cwf
    * product: plte
    * product: mbb
    * product: 5g mvc
* *Workstream/team*
    * workstream: devops_pe
    * workstream: reliability_be

#### Priority labes
* *Issue priority: applicable to bugs and improvements*
    * priority: critical
    * priority: high
    * priority: medium
    * priority: low
* *Execution order priority: useful for stack ranking execution priority*
    * priority: p0
    * priority: p1
    * priority: p2
    * priority: p3
#### Custom
* *Relevant custom labels*
    * good first issue
    * good intern project
    * Epic
    * breaking change
    * vX.Y
    * apply-vX.Y
    * affects-vX.Y
    * backported-vX.Y

**Comment by hcgatewood** [comment]
Also need to include the `backported-vX.Y` as we use that to indicate completion of the `apply-vX.Y` label

P.S. thanks for adding this @ekowtaylor, looks great

**Comment by hcgatewood** [comment]
@electronjoe will probably want some input here, I see a number of labels he created will get deleted, e.g. https://github.com/magma/magma/labels/source%3Aclang-tidy

**Reply by electronjoe** [reply]
Thanks Hunter! Yeah my use for some of these source: labels was to enable software engineers to track better engineering improvements and their impacts by label use.  E.g. at the end of the half to be able to link to label:clang-tidy and say: the following PRs all were a result of unblocking Clang Tidy work.  Or Clang compile, etc.

* source:clang-tidy
* source:clang

Are the two I have created so far I think.

I could see myself adding:

* source:tsan soon...

Do we think these pollute the label space too much? Github doesn't seem super great with alternative methods - but I suppose I could go edit Issues / PRs and add custom hashtags or something and query by text...

**Reply by electronjoe** [reply]
Also we have a set of proposed labels for Proposal processes - should they involve GH Issues (hopefully this will be finalized this week).

That then adds:

* `proposal:needs:design doc`
  * Proposal has been deemed to merit a more detailed design doc.
  * Please ensure it Addresses the design questions raised in this Github Issue.
* `proposal:status:In Review`
  * This proposal is in active review, all discussion will be within this Github Issue.
* `proposal:status:Accepted`
  * This proposal has been accepted by general consensus
* `proposal:status:Rejected`
  * This proposal has been rejected by general consensus

**Reply by ekowtaylor** [reply]
I think the engineering improvement tags are a great addition but I believe the proposals can still leverage the generic status tags. I added 'rejected'. What do you think?

**Comment by mattymo** [comment]
@ekowtaylor  Please don't leave out the functional use behind the two labels currently in use for LTE and CWAG tests: "component: agw" and "component: cwag". These are used to trigger LTE and CWAG integration tests in Jenkins. The other "component" tags don't have any special effects currently.

**Reply by ekowtaylor** [reply]
Will add them. Thanks.

**Comment by electronjoe** [comment]
I am finding labels with spaces in them non-ergonomic (you have to include the label with quotes when filtering in Github UIs).  Can we remove the spaces? How do people feel about that.

**Reply by mattymo** [reply]
> 
> 
> I am finding labels with spaces in them non-ergonomic (you have to include the label with quotes when filtering in Github UIs). Can we remove the spaces? How do people feel about that.

I like the aesthetics of labels with spaces. I do not find it difficult to quote labels in API calls

**Reply by electronjoe** [reply]
I was perhaps not thorough in my explanation of non ergonomic =D

* Newcomers do not know how to search for a label: and have to explore via the UI
  * I failed to filter by space-containing labels until I explored and reverse engineered the quotes
* Using a non-space delineator promotes hierarchical label use
  * E.g. component:agw:mme - versus "component: agw ?? MME" or flat label space for all the things
  * I find that better for managing labels if there are a lot of them
  * But perhaps having a lot of labels is not a good plan

2 cents

**Reply by electronjoe** [reply]
Here are examples of the Proposal process labels I had proposed - which use the hierarchical label use as an example:

* `proposal:needs:design doc`
  * Proposal has been deemed to merit a more detailed design doc.
  * Please ensure it Addresses the design questions raised in this Github Issue.
* `proposal:status:In Review`
  * This proposal is in active review, all discussion will be within this Github Issue.
* `proposal:status:Accepted`
  * This proposal has been accepted by general consensus
* `proposal:status:Rejected`
  * This proposal has been rejected by general consensus

Edit: But I am noticing that I used spaces in the final label endpoint... Hah!

**Reply by hcgatewood** [reply]
`proposal:status:in_review` ?
